Update conv creation to use prefill flags

justinrosner opened the pull request:
## Motivation

There is no need for rocMLIR to keep additional logic around for zero init kernels when we already have a cleaner way of implementing this with prefill attributes. The main goal of this PR is to consolidate these logic branches into one, and remove the redundant zero init kernel way of doing things.

## Technical Details

This PR does the following:
- Move ConvGenerator over to using prefill attributes, similar to what is done for split-k, instead of using explicit init kernels.
- Update logic in ConvToGemm in the non-V4R1 case for bwd_data conv ops

## Test Plan

- MLIR LIT Tests
- CIs are all Green 

## Test Result

- [x] MLIR LIT Tests
- [ ] CIs are all green

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

justinrosner committed:
Use prefill attributes instead of rock.init_kernel ops

justinrosner committed:
Fix typo in comment

justinrosner committed:
Small fixes and updates to LIT tests

justinrosner committed:
Clang-format

umangyadav commented:
@justinrosner  there are some merge conflicts.

umangyadav commented on mlir/include/mlir/Dialect/Rock/utility/fusionUtils.h:
```diff
@@ -30,6 +30,9 @@ LogicalResult testFusionLegalitySplitK(func::FuncOp func);
 // the atomic operation is supported by the hardware.
 LogicalResult testFusionLegalityReduce(func::FuncOp func);
 
+// Checks whether a function contains any `rock::ConvBwd*` ops
```
It is supposed to test fusion Legality not just presence of `rock::ConvBwd`

justinrosner replied:
In our earlier conversations I thought that you mentioned that we initially want to disable all fusions for backwards conv ops? Or is your comment more specifically just for updating the comment?

umangyadav commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -277,28 +277,17 @@ LogicalResult ConvGenerator::getBwdWeightKernelCount(OpBuilder &builder,
     }
     if (!needExtraPad) {
       Type dataType = getInputDataType(builder);
-      if (dataType.isF32()) {
```
Q: Why is Fp32 case removed here ?

dhernandez0 replied:
I think it's because the first kernel was 0-init and now that will be handled by prefill attribute? @justinrosner

justinrosner replied:
Yes, that is correct!

umangyadav commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -757,6 +746,24 @@ ConvolutionDims ConvGenerator::getConvolutionDims(const Config *config) {
 
+// Helper function to zero-initialize an argument of a FuncOp using the
+// prefill attribute
+void zeroInitArg(OpBuilder &builder, func::FuncOp func, unsigned argIndex) {
```
Use `static void` if not used outside this cpp file.

umangyadav commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -954,41 +948,47 @@ LogicalResult ConvGenerator::genConvModule(ModuleOp &module, int rawKernelId,
+          needsZeroInit = true;
+        } else if (dataType.isF16()) {
+          needsZeroInit = true;
+        }
```
```suggestion
        if (dataType.isF32() || dataType.isF16()) {
          needsZeroInit = true;
        } 
```

dhernandez0 replied:
needsZeroInit = dataType.isF32() || dataType.isF16()

umangyadav commented on mlir/lib/Dialect/Rock/utility/fusionUtils.cpp:
```diff
@@ -231,3 +231,24 @@ LogicalResult mlir::rock::testFusionLegalityReduce(ModuleOp mod) {
+        return WalkResult::advance();
+      });
+
+  return success(!walkResult.wasInterrupted());
```
```suggestion
  return !walkResult.wasInterrupted();
```

It will get converted to success implictly.

justinrosner replied:
I don't think it will get converted implicitly. `LogicalResult` does not have a public constructor that takes a bool, so we get a compile error when trying to just return the bool.

umangyadav commented on mlir/lib/Dialect/Rock/utility/fusionUtils.cpp:
```diff
@@ -231,3 +231,24 @@ LogicalResult mlir::rock::testFusionLegalityReduce(ModuleOp mod) {
+
+LogicalResult mlir::rock::testFusionLegalityBwdConv(ModuleOp mod) {
+  auto funcs = mod.getOps<func::FuncOp>();
+  assert(std::distance(funcs.begin(), funcs.end()) &&
```
Address copilot comment

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h:
```diff
@@ -115,6 +115,13 @@ LogicalResult calculateKBlockNum(const int64_t batchSize,
                                  int64_t &nKBlock);
 
+/// This is a helper function for BwdData convolution ops to determine if every
+/// pixel is being written to, which ultimately determines if there will need
```
nit: pixel -> element? pixel only makes sense for 2D images IMO, not for features.

dhernandez0 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -277,28 +277,17 @@ LogicalResult ConvGenerator::getBwdWeightKernelCount(OpBuilder &builder,
+        // The first kernel will conduct the actual backward weight
+        // convolution, using atomic add instructions. The second kernel will do
         // elementwise conversion from fp32 in the workspace to fp16 in the
         // actual output (filter tensor).
```
we still do this second kernel internally, but migraphx will do it themselves when we integrate this with them, right?

justinrosner replied:
The second kernel that does the actual computation is now the first kernel (we are not getting rid of it here). It's just renumbered since we no longer need the init kernel.

justinrosner replied:
Wait, if you are talking about the kernel that does the workspace conversion then I'm not sure. I'll have to talk with ZhiRun about this.

dhernandez0 replied:
I think the idea was that migraphx would do that one as well. Because otherwise we need to change our API migraphx<->rocmlir to handle more than one kernel. And it's a trivial kernel anyway.

justinrosner replied:
See thread below. This will now just be used in rocmlir-gen as we are not creating multiple kernels for the MIGraphX flow

dhernandez0 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -954,41 +948,47 @@ LogicalResult ConvGenerator::genConvModule(ModuleOp &module, int rawKernelId,
+      // For all backwards data convolution ops that don't write to every pixel,
+      // we want to zeroinitialize the buffer in the second argument
+      // (input tensor)
+      zeroInitArg(builder, func, 1);
```
This is ok for now if we don't enable fusions. However, if we enable fusions I don't think there's a guarantee arg 1 is going to be the input tensor. Please add a TODO re this.

dhernandez0 replied:
wait, I guess this is used by rocmlir-gen only, right? then it's fine

justinrosner replied:
This is only used in rocmlir-gen right now, but in https://github.com/ROCm/rocMLIR/pull/1951 we are going to be using this in the TosaToRock conversion as well. So I'm going to go ahead with the TODO for the future when we want to handle fusions.

dhernandez0 replied:
right but you won't use ConvGenerator.cpp for TosaToRock, because ConvGenerator is for rock dialect only. So, IMO we don't need the TODO. What do you think?

justinrosner replied:
Could we not use an instance of ConvGenerator to deal with the creation of the multiple kernels that are needed for the backwards convolutions?

dhernandez0 replied:
yes. But I think we didn't want to have multiple kernels AFAIK (only internally for rocmlir-gen). I think the second kernel was supposed to be done by migraphx.

justinrosner replied:
That may be true for bwd_weight kernels, but I don't think that is the case for bwd_data kernels. We will potentially need to have multiple bwd_data kernels to handle all of the computation

dhernandez0 replied:
ok, that's good to know. That means, we need to also change the API with migraphx in order to support multiple kernels.

justinrosner replied:
I'll start a discussion with some MIGraphX folks tomorrow to get an answer for this

justinrosner replied:
We had some discussions with MIGraphX folks offline and they do not need the multiple kernels that rocMLIR currently generates, therefore the original assumption that Daniel made (about this only being used from rocMLIR) holds true.

dhernandez0 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -954,41 +948,47 @@ LogicalResult ConvGenerator::genConvModule(ModuleOp &module, int rawKernelId,
   case ConvOpType::BwdData: {
-    if (kernelId < 0) {
-      // zero init input tensor
-      builder.create<InitKernelOp>(
```
can we remove InitKernelOp (and any lowering related to it) if it's not used anywhere?

justinrosner replied:
Yes, removed.

dhernandez0 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -954,41 +948,47 @@ LogicalResult ConvGenerator::genConvModule(ModuleOp &module, int rawKernelId,
+      }
+    }
+
+    if (rawKernelId == 1) {
```
isn't rawKernelId always 0 now? Where is rawKernelId defined?

justinrosner replied:
Not quite. This is passed in as a function argument to `genConvModule`. This will be directly used when setting the `kernel` attribute for backwards data rock ops.

We can likely rename this to just kernelID now, since we no longer have to do any conversion on this value. I'll make that change now.

dhernandez0 commented on mlir/test/rocmlir-gen/kernel-repeats.mlir:
```diff
@@ -11,27 +11,23 @@
 // CONV_WRW_GPU-LABEL: func.func @rock_conv_bwd_weight_gk01c_n01gc_n01gk_0
-// CONV_WRW_GPU: rock.init_kernel
-// CONV_WRW_GPU-LABEL: func.func @rock_conv_bwd_weight_gk01c_n01gc_n01gk_1
+// CONV_WRW_GPU-SAME: %arg0: memref<262144xf32> {rock.prefill = 0.000000e+00 : f32}
```
all prefill tests seem to be f32, can we add f16 as well?

justinrosner replied:
I updated populate_bwd.mlir to have one of the tests check prefill for f16

justinrosner committed:
Attend to more review comments

justinrosner committed:
Remove InitKernel from Rock Ops

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h:
```diff
@@ -113,6 +113,13 @@ LogicalResult calculateKBlockNum(const int64_t batchSize,
+/// This is a helper function for BwdData convolution ops to determine if every
+/// element is being written to, which ultimately determines if there will need
+/// to be multiple kernels to handle the backward data convolution.
+bool isEveryPixelWritten(ArrayRef<int64_t> strideDims,
```
nit: Pixel -> Element ?

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/loweringUtils.cpp:
```diff
@@ -103,6 +103,20 @@ LogicalResult mlir::rock::calculateKBlockNum(const int64_t batchSize,
 
+// Heuristic to determine if every pixel in the output would be written by the
+// backward data convolution algorithm.
+bool mlir::rock::isEveryPixelWritten(ArrayRef<int64_t> strideDims,
```
nit: this seems very specific to ConvGenerator, do we want it here?

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/loweringUtils.cpp:
```diff
@@ -103,6 +103,20 @@ LogicalResult mlir::rock::calculateKBlockNum(const int64_t batchSize,
   return success();
 }
 
+// Heuristic to determine if every pixel in the output would be written by the
```
no need to have a description here if it's already in loweringUtils.h

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/loweringUtils.cpp:
```diff
@@ -116,25 +130,9 @@ mlir::rock::backwardDataKernelIds(ArrayRef<int64_t> strideDims,
-    }
-    return result;
-  };
-  bool needZeroInitKernel = !isEveryPixelWritten();
```
why don't we do this here anymore?

justinrosner replied:
In the population of kernelIDs we no longer need to explicitly check to see if every element is written as we will never have an init kernel. Instead, we have this same check when deciding on if we need to add prefill attributes elsewhere in ConvGenerator

justinrosner committed:
Move and rename isEveryElementWritten

dhernandez0 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -955,41 +964,46 @@ LogicalResult ConvGenerator::genConvModule(ModuleOp &module, int rawKernelId,
           /*elemsPerThread=*/nullptr);
     } else {
+      if (needsZeroInit) {
+        zeroInitArg(builder, func, hasWorkspace ? 3 : 0);
```
the TODO applies to this line as well. Can we add a TODO here as well?

justinrosner committed:
Add additional comment

justinrosner committed:
git clang-format

justinrosner committed:
Fix rock namespace

umangyadav commented:
@justinrosner  External CI failing. Please check

justinrosner commented:
> @justinrosner External CI failing. Please check

Talking with CI team to try and get a working machine. Was running into difficulties last week.

justinrosner committed:
Add fusability test

dhernandez0 commented on mlir/test/fusion/fusability-bwd-conv.mlir:
```diff
@@ -0,0 +1,35 @@
+    memref.copy %7, %arg4 : memref<256x1x64x112x112xf32> to memref<256x1x64x112x112xf32>  
+    return  
+  }  
+}
```
new line

dhernandez0 commented:
on second thought. I think conv bacward should be fusible if v4r1 is disabled, right? in that case there's no atomic add? @justinrosner

justinrosner committed:
Add newline

justinrosner commented:
> on second thought. I think conv bacward should be fusible if v4r1 is disabled, right? in that case there's no atomic add? @justinrosner

@umangyadav If I recall correctly, you initially had some concerns that this should be left for a separate PR, after there had been some discussions with the MIGraphX team. Now that we have split up the implementation and have v4r1 and non-v4r1 versions, is there any reason why we wouldn't want to fuse non-v4r1 bwd_conv ops? I.e., bwd_data convs that are implemented entirely in a single kernel

justinrosner commented:
> > on second thought. I think conv bacward should be fusible if v4r1 is disabled, right? in that case there's no atomic add? @justinrosner
> 
> @umangyadav If I recall correctly, you initially had some concerns that this should be left for a separate PR, after there had been some discussions with the MIGraphX team. Now that we have split up the implementation and have v4r1 and non-v4r1 versions, is there any reason why we wouldn't want to fuse non-v4r1 bwd_conv ops? I.e., bwd_data convs that are implemented entirely in a single kernel

@dhernandez0 Actually, the more that I think about this, do we even need specific fusion checks for bwd_data at all? We already have atomic checks that will run regardless of the op (which will cover the case when there is atomic add)

dhernandez0 commented:
> > > on second thought. I think conv bacward should be fusible if v4r1 is disabled, right? in that case there's no atomic add? @justinrosner
> > 
> > 
> > @umangyadav If I recall correctly, you initially had some concerns that this should be left for a separate PR, after there had been some discussions with the MIGraphX team. Now that we have split up the implementation and have v4r1 and non-v4r1 versions, is there any reason why we wouldn't want to fuse non-v4r1 bwd_conv ops? I.e., bwd_data convs that are implemented entirely in a single kernel
> 
> @dhernandez0 Actually, the more that I think about this, do we even need specific fusion checks for bwd_data at all? We already have atomic checks that will run regardless of the op (which will cover the case when there is atomic add)

I think as long as we have a special function (testFusionLegalityBwdConv) for backward fusions, we want to test it works correctly.
AFAIK the atomic checks only run if split-k is enabled (perf_config.splitk > 1).

IMO, we can make testFusionLegalityBwdConv() as simple as: fusability = !v4r1. The split-k function (testFusionLegalitySplitK) can handle the split-k case for backward as well I think. Then, we just need a test with fusion, returning fusability=true (if v4r1=false, and fusability=false if v4r1=true) and some split-k+backward tests just in case.

justinrosner committed:
Remove fusibility checks

justinrosner commented:
> > > > on second thought. I think conv bacward should be fusible if v4r1 is disabled, right? in that case there's no atomic add? @justinrosner
> > > 
> > > 
> > > @umangyadav If I recall correctly, you initially had some concerns that this should be left for a separate PR, after there had been some discussions with the MIGraphX team. Now that we have split up the implementation and have v4r1 and non-v4r1 versions, is there any reason why we wouldn't want to fuse non-v4r1 bwd_conv ops? I.e., bwd_data convs that are implemented entirely in a single kernel
> > 
> > 
> > @dhernandez0 Actually, the more that I think about this, do we even need specific fusion checks for bwd_data at all? We already have atomic checks that will run regardless of the op (which will cover the case when there is atomic add)
> 
> I think as long as we have a special function (testFusionLegalityBwdConv) for backward fusions, we want to test it works correctly. AFAIK the atomic checks only run if split-k is enabled (perf_config.splitk > 1).
> 
> IMO, we can make testFusionLegalityBwdConv() as simple as: fusability = !v4r1. The split-k function (testFusionLegalitySplitK) can handle the split-k case for backward as well I think. Then, we just need a test with fusion, returning fusability=true (if v4r1=false, and fusability=false if v4r1=true) and some split-k+backward tests just in case.

I've moved the fusibility checks over to https://github.com/ROCm/rocMLIR/pull/1951 so that we can make use of the `usesV4R1` attribute. This PR now just focuses on the prefill attributes

justinrosner committed:
Fix rebase issues and make updates for bwd_data

umangyadav committed:
fix formatting

umangyadav committed:
Merge branch 'develop' into justinr-prefill-and-fusability
